int32 tf_x
int32 tf_y
int32 tf_z

